---
title: "Data QC report"
date: "`r format(Sys.time(), '%d/%m/%y')`"
output: 
  html_document:
    code_folding: hide
    toc: true
params:
    dataset_name: "unknown"
    platform: "unknown"
    N: 100
    S: 0.4
    SD: 3
    SD_exp: 4
---

# Dataset

This report is for cohort: **`r params$dataset_name`**.

Expression platform for this cohort is: **`r params$platform`**.

Post-QC sample size for this cohort is: **`r library(data.table); exp <- fread(params$N); nrow(exp)`**

**Settings used for sample outlier detection:**

Used LOF S-threhold for genetic outliers: `r params$S`

Threshold of SD difference from mean genetic PC1 and PC2 is: `r params$SD`

Threshold of SD difference from mean gene expression PC1 and PC2 is: `r params$SD_exp`

## Genotype QC report

### Overview of the QC steps

```{r, message = FALSE, warning = FALSE, echo = FALSE}
library(data.table)
library(knitr)
library(ggplot2)

overview_table <- fread("outputfolder_gen/gen_data_summary/summary_table.txt")
knitr::kable(overview_table)
```

### Sex check

```{r conditional_check_sex, eval=file.exists("outputfolder_gen/gen_plots/SexCheck.png")}
print("Next plot summarises the genetically determined sex of the input samples. Samples which have unclear genetic sex (X chr heterozygosity >0.2 and <0.8) and hence might be contaminated, are removed. If reported sex is indicated in the input `.fam` file, this is also compared with genetic sex and problematic samples (potential mixups) are automatically excluded.

You can investigate the file `[your output folder]/outputfolder_gen/gen_data_QCd/SexCheck.txt` to see which samples failed the sex check.

![**Fig 1. Distribution of F-values representing X chromosome heterozygosity.** Genetically determined sex thresholds are shown as vertical lines. Samples with the F-value between those two lines are potentially contaminated samples and are automatically excluded.](outputfolder_gen/gen_plots/SexCheck.png)")
```
```{r conditional_no_x, eval=!file.exists("outputfolder_gen/gen_plots/SexCheck.png")}
print("Step skipped since no X chromosome was found in the genotype data")
```

### Excess heterozygosity

Next plot summarises the heterozygosity of the input samples. Samples showing excess/depleted heterozygosity rate (mean+/-3SD) have potential issues with contamination or inbreeding, and are automatically excluded.

![**Fig 2. Distribution of heterozygosity rate.** Thresholds for declaring excess/depleted heterozygosity for sample (mean+/-3SD) are shown as vertical lines. ](outputfolder_gen/gen_plots/HetCheck.png)  

### Ancestry outliers

#### Genotype data projected into 1000G superpopulations

Here we project your genotype samples into 1000G p3v5 superpopulations to check if all samples come from one ancestry. 

1. If you observe a few samples which are different ancestry compared to the majority of samples (e.g. majority of the samples are of EUR ancestry and there are few samples from AFR or SAS), then remove ancestry outliers and re-run the DataQc pipeline without those samples.
2. If you observe many (>100) samples which are different ancestry compared to the majority of samples (your dataset consists of individuals from several ancestries) then please split your genotype data based on ancestry and rerun this and following pipelines on those datasets separately.

The file `AssignedPopulations.txt` and `1000G_PC_projections.txt` has been written to the your output folder under `gen_data_summary`. These can be used to extract the outlier samples based on their PC projections into 1000G superpopulations. Easiest is to make another genotype-to-expression (gte) file which has those samples removed and rerun the pipeline with new gte. You can also use plink/plink 1.9/plink 2.0 to filter samples from initial .bed/.bim/.fam files.

![**Fig 7. Genotype samples projected to 1000G superpopulations.** First two plots show only target dataset in the 1000G PC space. Rest of the plots visualise target dataset, as well as samples from 1000G superpopulations by plotting first 10 genotype PCs.](outputfolder_gen/gen_plots/SamplesPCsProjectedTo1000G.png)  

#### Assign superpopulation for each sample

Based on first 3 genotype PCs, here we assign the most likely population to each sample and outline those samples which are clearly different. Again, all the samples in the dataset should be the most similar to only one ancestry and samples resembling other ancestries should be removed or treated as a separate dataset (in case of N>100).

```{r, fig.height = 7, fig.width = 7, fig.align = "center", echo = FALSE, message = FALSE, warning = FALSE}
library(dplyr)
pops <- fread("outputfolder_gen/gen_data_summary/PopAssignResults.txt")

pops <- melt(pops)
colnames(pops) <- c("Sample", "Superpopulation", "dist") 

pops <- pops %>% 
group_by(Sample) %>% 
arrange(dist, .by_group = TRUE) %>% 
mutate(SamplePopAssign = head(Superpopulation, 1)) %>%
mutate(Difference = dist[2]/dist[1]) %>%
mutate(Confidence = case_when(Difference > 2 ~ "confident", Difference <= 2 ~ "not confident"))

res <- unique(pops[, c(1, 4, 5, 6)])

# Visualise
ggplot(pops, aes(x = Superpopulation, y = dist, fill = Superpopulation)) + 
geom_point() + 
geom_violin(draw_quantiles = 0.5) + 
theme_bw() + 
scale_fill_manual(values = c("Target" = "black", 
"EUR" = "blue", 
"EAS" = "goldenrod", 
"AMR" = "lightgrey", 
"SAS" = "orange", 
"AFR" = "red"))

fwrite(res, "outputfolder_gen/gen_data_summary/AssignedPopulations.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

**Fig 6. Euclidean distance between target samples and samples from each 1000G superpopulation.** 

`r nrow(res[res$SamplePopAssign == "EUR",])` samples in the data are most similar to EUR superpopulation.

`r nrow(res[res$SamplePopAssign == "AMR",])` samples in the data are most similar to AMR superpopulation.

`r nrow(res[res$SamplePopAssign == "EAS",])` samples in the data are most similar to EAS superpopulation.

`r nrow(res[res$SamplePopAssign == "SAS",])` samples in the data are most similar to SAS superpopulation.

`r nrow(res[res$SamplePopAssign == "AFR",])` samples in the data are most similar to AFR superpopulation.

`r res %>% filter(Confidence != "confident") %>% nrow()` samples assignement to the data is not very confident (<2 times difference with second most similar superpopulation).

Again, please follow the following instructions:

1. If you observe a few samples which are different ancestry compared to the majority of samples (e.g. majority of the samples are of EUR ancestry and there are few samples from AFR or SAS), then remove ancestry outliers and re-run the DataQc pipeline without those samples.
2. If you observe many (>100) samples which are different ancestry compared to the majority of samples (your dataset consists of individuals from several ancestries) then please split your genotype data, based on ancestry and rerun this and following pipelines on those datasets separately. This assumes that you know the ancestry information for your genotype data.

The file `AssignedPopulations.txt` and `1000G_PC_projections.txt` has been written to the your output folder under `gen_data_summary`. These can be used to extract the outlier samples based on their PC projections into 1000G superpopulations. Easiest is to make another genotype-to-expression (gte) file which has those samples removed and rerun the pipeline with new gte. You can also use plink/plink 1.9/plink 2.0 to filter samples from initial .bed/.bim/.fam files.


### Individual genetic outliers in the target population

Here we conduct PCA on the genotypes of eQTL samples only and investigate if there are any individual outliers. These are samples which have the same ancestry as bulk of the samples, however still show somewhat different genetic makeup. This might be due to considerable population stratification (different populations or ethnicities in the same superpopulation) or due to technical reasons. By default, we remove clear outliers.

#### Outlier samples

On the next plot you can see the distribution of Probabilistic Local Outlier Factor "outlierness" statistic ([Kriegel et al., 2009](https://dl.acm.org/doi/10.1145/1645953.1646195), [PrivÃ© et al., 2020](https://academic.oup.com/bioinformatics/article/36/16/4449/5838185)), based on which the samples are declared genetic outliers and removed from the analysis. You should check based on the histogram if the default value is appropriate. If not, adjust it in the pipeline argument and re-run.

If you observe bimodal distribution on the histogram then it might indicate substratification in the genotype data. Check the following section, Figure 4 and 5. 

![**Fig 3. Statistic of outlierness with threshold used to remove ethnic outliers.**](outputfolder_gen/gen_plots/PC_dist_outliers_S.png)  

On the next two plots we visualise genetic principal components to check if genetic outliers were removed. If you observe unsolved sub-clusters on following PCA plots, this might indicate that you have different genotyping batches in the data, or large fraction of samples (e.g. >10%) from somewhat different ancestry. In this case, the algorithm migh not assign all deviating samples as outliers or remove unreasonably large fraction of samples. 

If the pipline removes unreasonably large fraction of the samples, there are multiple courses of action to consider:

1. If you observe clusters with limited numbers of clusters (e.g. N<500) and a clear separation based on PC1 or PC2, you can consider using the data as it is: first 2 genetic PCs will be used as covariates in the meta-analysis and such batch effects will be mostly accounted for. In that case, you should re-run the `DataQC` pipeline with large SD value, e.g. `--SDthresh 10`, to keep all the data in.

2. If you have multiple batches with many samples (N>500), you might consider splitting the data based on genotyping batch and re-running `DataQC` and all following pipelines on those batches.

If the pipelines does not remove all clear outlier samples, adjust `--SDthresh` threshold accordingly, to remove small fraction of clearly deviating samples.

In case of doubt please contact the lead analysts for advice and we will investigate further.

![**Fig 4. PCA with all samples, outliers outlined with red.** Different shades of red indicate different classes of outliers: difference in SD from mean of PC1 and PC2 (pink), statistic of outlierness S (red), or both (dark red).](outputfolder_gen/gen_plots/PCA_outliers.png)  

![**Fig 5. PCA with all samples, outliers removed.** You should not see any dramatic outliers any more. These are the PCs which will be used as covariates in the eQTL meta-analysis.](outputfolder_gen/gen_plots/Target_PCs_postQC.png)  

#### Loadings of PCs

Visualise the loadings of 10 first PCs calculated on final quality-controlled genotype data. Because PC calculation command removed long-range LD, you should not see any peaks on the plots any more, those would indicate that PC is still driven by LD.

![**Fig 6. Loadings of 10 first PCs.**](outputfolder_gen/gen_plots/Target_PCs_postQC_Loadings.png)

## Gene expression QC report

### Overview of the QC steps

```{r, message = FALSE, warning = FALSE, echo = FALSE}
overview_table <- fread("outputfolder_exp/exp_data_summary/summary_table.txt")
knitr::kable(overview_table)
```

### PCA on preprocessed expression data

On the following plots, 8 first PCs are calculated on normalised and log~2~-transformed expression matrix. Samples were declared as outliers based on first two PCs, when they deviate from the mean value by `r params$SD_exp` standard deviations (SDs). Analysis was run iteratively, data was pre-processed and PCA was calculated until there were no further expression otliers to remove. Because there might be expected difference based on the sex, we also indicate the genetic sex of each sample on the plot.

![**Fig 8. PCA on all samples** Genetic sex of the individual is shown as shape of the dot, outliers from all iterations are outlined as red.](outputfolder_exp/exp_plots/PCA_before.png)

In order to visualise non-linear patterns in the data, we calculated 2 first multidimensional scaling (MDS) components and visualise those as well. This analysis were done on the data where outliers from previous step were already removed. Similarly to PCs, samples were declared as outliers when they deviate from the mean value of MDS 1 and MDS 2 by `r params$SD_exp` SDs. If you still observe outlier groups on this plot, you should try to reduce the SDs allowed to deviate (parameter `--ExpSdThreshold` of the pipeline) and re-run the pipeline.

![**Fig 9. MDS on all samples** Genetic sex of the individual is shown as shape of the dot, outliers are outlined as red.](outputfolder_exp/exp_plots/MDS_before.png)

These are 8 first expression PCs, calculated on fully processed expression data (normalised and INT-transformed) after removal of outlier samples. Genetic sex of the individual is shown as shape of the dot. You should not see any dramatic outliers any more on first two PCs.

![**Fig 10. PCA on all samples, outliers removed.** Genetic sex of the individual is shown as shape of the dot.](outputfolder_exp/exp_plots/PCA_after.png) 

#### Explained variance of expression PCs

![**Fig 11. PCA scree plot for normalised and inverse normal transformed expression matrix.** Outlier samples were removed before re-calculating PCA.](outputfolder_exp/exp_plots/PCA_final_scree_plot.png)

Explained variance of final 100 first PCs which are used as covariates in the analysis (normalised and inverse normal transformed expression matrix).

```{r, message = FALSE, warning = FALSE, fig.width = 14, fig.height = 4, fig.align = "center"}
overview_pcs <- fread("outputfolder_exp/exp_data_summary/summary_pcs.txt")

overview_pcs_cumulative <- overview_pcs

overview_pcs_cumulative$cumulative <- cumsum(overview_pcs$explained_variance)

overview_pcs_cumulative$PC <- factor(overview_pcs_cumulative$PC, levels = as.character(overview_pcs_cumulative$PC))

p <- ggplot(overview_pcs_cumulative, aes(x = PC, y = cumulative)) + 
geom_bar(stat = "identity") + 
theme_bw() + 
scale_y_continuous(limits = c(0, 1)) +
theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

p
```

**Fig 12. Cumulative explained variance for first 100 PCs.**

First 5 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[5] * 100`% of variance.

First 10 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[10] * 100`% of variance.

First 20 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[20] * 100`% of variance.

First 50 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[50] * 100`% of variance.

First 100 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[100] * 100`% of variance.

```{r, message = FALSE, warning = FALSE}
knitr::kable(overview_pcs)
```

### Sex-based mixup check

Next plot shows the expression of the sex chromosomes genes. We assume that males have higher expression of genes encoded from Y-chromosome and females have higher expression of X chromosome gene *XIST*. Therefore you should see that subsets of samples align parallel to x-axis (females) and y-axis (males). There should be no samples on the x-y diagonal. 

Samples which expression does not align with genetic sex are potential sample mixups. These are outlined with red and automatically removed from further analyses.

![**Fig 13. Expression of X/Y chromosome genes.** On the x-axis is the normalised expression level of XIST, on the y-axis is the mean of the normalised expression levels of several Y chromosome genes. Point type indicates the genetic sex and there should be no samples on the X-Y diagonal. Samples which are in the wrong sample cloud, are likely sample mixups and were automatically removed from the data.](outputfolder_exp/exp_plots/SexSpecificGenes.png) 
