---
title: "Data QC report"
date: "`r format(Sys.time(), '%d/%m/%y')`"
output: 
  html_document:
    code_folding: hide
params:
    dataset_name: "unknown"
    platform: "unknown"
---

# Dataset

This report is for cohort: **`r params$dataset_name`**.

Expression platform for this cohort is: **`r params$platform`**.

## Genotype QC report

### Overview of the QC steps

```{r, message = FALSE, warning = FALSE, echo = FALSE}
library(data.table)
library(knitr)
library(ggplot2)

overview_table <- fread("outputfolder_gen/gen_data_summary/summary_table.txt")
knitr::kable(overview_table)
```

### PCA on target data

#### Outlier samples

On the next two plots you can see the "outlierness" statistic based on which the samples are declared genetic outliers and removed from the analysis. You should check based on the histogram if the default value is appropriate. If not, adjust it in the pipeline argument and re-run.

If you observe bimodal distribution, this could indicate that you have substantial number of samples from multiple ancestries in your data. If you have >100 samples from either anchestries, you should split the data and re-run the current and following pipelines on those subsets. This assumes that you know this information about your genotype data: in case of doubt please contact urmo.vosa@gmail.com for advice and we will investigate further.

![**Fig 1. Statistic of outlierness with threshold used to remove ethnic outliers.**](outputfolder_gen/gen_plots/PC_dist_outliers_S.png)  

Here we visualise genetic principal components to check if genetic outliers were removed.

![**Fig 2. PCA with all samples, outliers outlined with red.**](outputfolder_gen/gen_plots/PCA_outliers.png)  

![**Fig 3. PCA with all samples, outliers removed.**](outputfolder_gen/gen_plots/Target_PCs_postQC.png)  

#### Loadings of PCs

Visualise the loadings of 10 first PCs. Because PC calculation command removed long-range LD, you should not see any peaks on the plots any more, those would indicate that PC is still driven by LD.

![**Fig 4. Loadings of 10 first PCs.**](outputfolder_gen/gen_plots/Target_PCs_postQC_Loadings.png)

#### Genotype data projected to 1000G superpopulations

Here we project your genotype samples into 1000G p3v5 superpopulations to check if all samples come from one ethnicity. 

1. If you observe a few samples which are different ancestry compared to the majority of samples (e.g. majority of the samples are of EUR ancestry and there are few samples from AFR or SAS), then remove ethnic outliers and re-run the GenotypeQc pipeline without those samples.
2. If you observe many (>100) samples which are different ancestry compared to the majority of samples (your dataset consists of individuals from several ancestries) then please split your genotype data based on ancestry and rerun this and following pipelines on those datasets separately.

The file `1000G_PC_projections.txt` has been written to the your output folder under `gen_data_summary`. This can be used to manually extract the outlier samples based on their PC projections into 1000G superpopulations. You can use plink/plink 1.9/plink 2.0 to filter samples from initial .bed/.bim/.fam files.

![**Fig 5. Genotype samples projected to 1000G superpopulations.**](outputfolder_gen/gen_plots/SamplesPCsProjectedTo1000G.png)  

#### Assign population for each sample

Based on first 3 genotype PCs, assign the most likely population to each sample, outline those samples which are clearly different.

```{r, fig.height = 7, fig.width = 7, fig.align = "center", echo = FALSE, message = FALSE, warning = FALSE}
library(dplyr)
pops <- fread("outputfolder_gen/gen_data_summary/PopAssignResults.txt")

pops <- melt(pops)
colnames(pops) <- c("Sample", "Superpopulation", "dist") 

pops <- pops %>% 
group_by(Sample) %>% 
arrange(dist, .by_group = TRUE) %>% 
mutate(SamplePopAssign = head(Superpopulation, 1)) %>%
mutate(Difference = dist[2]/dist[1]) %>%
mutate(Confidence = case_when(Difference > 2 ~ "confident", Difference <= 2 ~ "not confident"))

res <- unique(pops[, c(1, 4, 5, 6)])

# Visualise
ggplot(pops, aes(x = Superpopulation, y = dist, fill = Superpopulation)) + 
geom_point() + 
geom_violin(draw_quantiles = 0.5) + 
theme_bw() + 
scale_fill_manual(values = c("Target" = "black", 
"EUR" = "blue", 
"EAS" = "goldenrod", 
"AMR" = "lightgrey", 
"SAS" = "orange", 
"AFR" = "red"))
```

**Fig 6. Euclidean distance between target samples and samples from each 1000G superpopulation.** 


`r nrow(res[res$SamplePopAssign == "EUR",])` samples in the data are most similar to EUR superpopulation.

`r nrow(res[res$SamplePopAssign == "AMR",])` samples in the data are most similar to AMR superpopulation.

`r nrow(res[res$SamplePopAssign == "EAS",])` samples in the data are most similar to EAS superpopulation.

`r nrow(res[res$SamplePopAssign == "SAS",])` samples in the data are most similar to SAS superpopulation.

`r nrow(res[res$SamplePopAssign == "AFR",])` samples in the data are most similar to AFR superpopulation.

`r res %>% filter(Confidence != "confident") %>% nrow()` samples assignement to the data is not very confident (<2 times difference with second most similar superpopulation).

## Gene expression QC report

### Overview of the QC steps

```{r, message = FALSE, warning = FALSE, echo = FALSE}
overview_table <- fread("outputfolder_exp/exp_data_summary/summary_table.txt")
knitr::kable(overview_table)
```

### PCA on preprocessed expression data

On the following plots, 8 first PCs are calculated on normalised and log~2~-transformed expression matrix. Samples are declared as outliers based on first two PCs, when they deviate from the mean value by 4 standard deviations (SDs). Analysis was run iteratively, data was pre-processed and PCA calculated until there were no further expression otliers to remove.

![**Fig 7. PCA on all samples, outliers from first iteration outlined.**](outputfolder_exp/exp_plots/PCA_before.png)

These are 8 first expression PCs, calculated on fully processed expression data (normalised and INT-transformed) after removal of outlier samples. You should not see any dramatic outliers any more.

![**Fig 8. PCA on all samples, outliers removed.**](outputfolder_exp/exp_plots/PCA_after.png) 

#### Explained variance of expression PCs

![**Fig 9. PCA scree plot for normalised and inverse normal transformed expression matrix.** Outlier samples were removed before re-calculating PCA.](outputfolder_exp/exp_plots/PCA_final_scree_plot.png)

Explained variance of final 50 PCs which are used as covariates in the analysis (normalised and inverse normal transformed expression matrix).

```{r, message = FALSE, warning = FALSE, fig.width = 14, fig.height = 4, fig.align = "center"}
overview_pcs <- fread("outputfolder_exp/exp_data_summary/summary_pcs.txt")

overview_pcs_cumulative <- overview_pcs

overview_pcs_cumulative$cumulative <- cumsum(overview_pcs$explained_variance)

overview_pcs_cumulative$PC <- factor(overview_pcs_cumulative$PC, levels = as.character(overview_pcs_cumulative$PC))

p <- ggplot(overview_pcs_cumulative, aes(x = PC, y = cumulative)) + 
geom_bar(stat = "identity") + 
theme_bw() + 
scale_y_continuous(limits = c(0, 1)) +
theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

p
```

**Fig 10. Cumulative explained variance for first 100 PCs.**


First 5 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[5] * 100`% of variance.

First 10 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[10] * 100`% of variance.

First 20 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[20] * 100`% of variance.

First 50 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[50] * 100`% of variance.

First 100 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[100] * 100`% of variance.

```{r, message = FALSE, warning = FALSE}
knitr::kable(overview_pcs)
```
